### PR TITLE
Add test_example_59

### PR DESCRIPTION
**Pull request by kkloberdanz**
CDRIVER-4295

related: DRIVERS-2181

Implement a snapshot query example for the manual.

### CODE REVIEW

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:2563]: To apply a `mongoc_client_session_t` to an operation, it needs to be appended to the options with `mongoc_client_session_append`.

```suggestion
   bson_t opts;

   if (!mongoc_client_session_append (cs, &opts, &error)) {
      MONGOC_ERROR ("could not apply session options");
      goto cleanup;
   }
   cursor = mongoc_collection_aggregate (
      collection, MONGOC_QUERY_NONE, pipeline, &opts, NULL);
   bson_destroy (&opts);
```

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:2589]: Suggest using `bson_iter_as_int64` to handle multiple possible numeric types.
```suggestion
      *accumulator += bson_iter_as_int64 (&iter);
```

**Comment by kevinAlbs** [comment]
Looking good. Include CDRIVER-4295 in the commit message before merging. That will result in a comment with the commit being added to the Jira ticket.

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (galon1) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:2638]: It might be helpful to add a comment here describing that we are using an aggregation pipeline to return the count, since other drivers have been displaying the aggregation pipeline command used. Since this code example is already longer than other drivers, I think a comment is ok.

**Review by galon1 — APPROVED** [review]
LGTM!

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kkloberdanz) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:2638]: I'm glad you brought this up, because that's something I'm curious about. It would be a fair amount of duplicate code and a bit lengthy to inline both calls to `accumulate_adoptable_count` in this function, but I'm wondering if I should do one of the following:

1. Include the entire function `accumulate_adoptable_count` in the example, i.e., place the start example comment just above the declaration of `accumulate_adoptable_count`, and end example comment stays where it is within `test_example_59`.
2. Place the text of `accumulate_adoptable_count` in a comment in `test_example_59`.
3. Explain in a comment what is happening in the function `accumulate_adoptable_count` purely in English, and omit code in the comment.
4. Inline all the code from `accumulate_adoptable_count` within `test_example_59`.

I'm leaning towards 1., but I'm not sure how that would look when rendered. I'd like to know what everyone thinks about this, since I'm not sure which way would be the cleanest for the purpose of producing an example.

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kkloberdanz) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:2649]: I'm not quite sure how that works. I with the `Post` section, I was following the style of other examples in this file that place the cleanup code within a `Post` section. From other C examples that I saw in the documentation, it seems that cleanup code is being excluded, but I'm not sure if that is truly what we want.

For example, here's a function from the examples that leaves out the cleanup code:
https://www.mongodb.com/docs/v4.4/core/read-isolation-consistency-recency/

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kkloberdanz) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:2638]: P.S. I'm planning on reaching out to Jeffrey Allen, the reporter for the original ticket, regarding this before merging.

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kkloberdanz) [reply src/libmongoc/tests/test-mongoc-sample-commands.c]: One note on this, some platforms define int64_t to be `long`, while others define it to be `long long`, so portable print specifiers get a little hairy. There is a portable macro `PRId64` that solves this, but the code is just a little ugly.

```C
   printf ("there are %" PRId64 " adoptable pets\n", adoptable_pets_count);
```
This macro would fix the following warning:
```
[2022/11/14 19:18:51.748] /data/mci/53a76e388c58b50897191ae23d61ac25/mongoc/src/libmongoc/tests/test-mongoc-sample-commands.c:2639:26: error: format '%lld' expects argument of type 'long long int', but argument 2 has type 'int64_t {aka long int}' [-Werror=format=]
[2022/11/14 19:18:51.748]     printf ("there are %lld adoptable pets\n", adoptable_pets_count);
[2022/11/14 19:18:51.748]                        ~~~^
[2022/11/14 19:18:51.748]                        %ld
```

Leaving the variable as a `long long` would also be portable since a `long long` is defined to be at least 64 bits. Which way would you prefer?

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kkloberdanz) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:2639]: I got some feed back from Jeffery Allen. It sounds like he's in favor of option 3.

(quote from Jeff)
>Hey Kyle, thanks for reaching out. So my recollection with that ticket is that the main goal is to emphasize the snapshot functionality in the example. Therefore, I'm inclined to say that we can exclude the agg pipeline from the example, and leave a comment explaining the aggregation. That way, users aren't too overwhelmed by a large example and they'll still be able to see the snapshot behavior in action

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kkloberdanz) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:2658]: Jeff has checked off on using this comment as a part of the documentation.

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (galon1) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:2661]: I agree option 3 would be the clearest, but I'm also not sure about how strict the formatting constraints are. Maybe the docs team could help make the choice.

Comment (galon1) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:2661]: In [this page](https://www.mongodb.com/docs/v6.0/core/transactions/) the C example is much longer than others. So it could potentially be ok.

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (galon1) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:2677]: I think if space is not an issue then it would be helpful to have it. [This page](https://www.mongodb.com/docs/v6.0/core/transactions/) for example does include it.

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (galon1) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:2661]: I just saw the addition of the comment! Ignore my comment above.

Comment (kkloberdanz) [reply src/libmongoc/tests/test-mongoc-sample-commands.c]: That's interesting. I do like when an example is explicit, since this example includes 2 functions, then I suppose no reason why I can't either. I'll make that change for now, and request another review from both you and Kevin to see what you think.

**Comment by kkloberdanz** [comment]
FYI, Gil brought up that another C driver example is lengthy and has multiple functions, so it seems that if I were to include multiple functions as a part of the example, that would fit with how things are currently shown. I pushed that in order for us to compare and decide which way we would prefer.

For reference, the example that Gil found: https://www.mongodb.com/docs/v6.0/core/transactions/

**Review by galon1 — COMMENTED** [review]
I think the example looks good! I just have one question about using python in this example.

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kevinAlbs) [reply src/libmongoc/tests/test-mongoc-sample-commands.c]: We use `int64_t` in the codebase. I think it is safe to assume users have stdint.h. I prefer `int64_t` since it matches `bson_iter_as_int64` return.

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:2560]:
```suggestion
                        BCON_BOOL (true),
```

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:2559]:
```suggestion
                       "adoptable",
```

`BCON_UTF8` is for string values (not keys).

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kevinAlbs) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:2661]: Defer to Jeff's judgement since this example is requested by the documentation team.

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kevinAlbs) [inline src/libmongoc/tests/test-mongoc-sample-commands.c:2546]:
```suggestion
   bson_t opts = BSON_INITIALIZER;
```

`mongoc_client_session_append` expects the output BSON to be initialized.

**Review by kevinAlbs — APPROVED** [review]
LGTM with small fixes.

File: src/libmongoc/tests/test-mongoc-sample-commands.c
Comment (kkloberdanz) [reply src/libmongoc/tests/test-mongoc-sample-commands.c:2643]: I was thinking that the Python looked enough like regular MQL mixed with pseudo code that one would be able to more clearly see what the aggregation pipeline is doing. Do you think it needs to be called out that this is from the Python example?

**Pull request merged**
